src/pu-command.c

```diff
@@ -0,0 +1,339 @@
+
+    g_free(context);
+}
+
+static gint
+pu_unichar_get_width(gunichar c)
+{
+    if (G_UNLIKELY(g_unichar_iszerowidth(c)))
+        return 0;
+
+    if (g_unichar_iswide(c))
+        return 2;
+
+    return 1;
+}
+
+static gint
+pu_utf8_strlen(const gchar *s)
+{
+    gint len = 0;
+
+    g_return_val_if_fail(s != NULL, 0);
+
+    while (*s) {
+        len += pu_unichar_get_width(g_utf8_get_char(s));
+        s = g_utf8_next_char(s);
+    }
+
+    return len;
+}
+
+static gchar *
+get_main_help(PuCommandContext *context)
+{
+    GString *string = NULL;
+    GString *entry_string = NULL;
+    PuCommandEntry *entry = NULL;
+    g_autofree gchar *orig_help = NULL;
+    g_autoptr(GRegex) regex = NULL;
+    g_autoptr(GMatchInfo) match_info = NULL;
+    g_autoptr(GString) match_string = NULL;
+    gsize max_len = 0;
+
+    g_return_val_if_fail(context != NULL, NULL);
+
+    /* Get help text of option context */
+    orig_help = g_option_context_get_help(context->option_context, TRUE, NULL);
+    string = g_string_new(NULL);
+
+    /* Search original help text for length of options including whitespace */
+    regex = g_regex_new("  -h, --help\\s+", 0, 0, NULL);
+    g_regex_match(regex, orig_help, 0, &match_info);
+    if (g_match_info_matches(match_info)) {
+        match_string = g_string_new(g_match_info_fetch(match_info, 0));
+        max_len = match_string->len;
+    } else {
+        max_len = 6 + pu_utf8_strlen(context->entries[0].name);
+    }
+
+    /* Construct command help text */
+    g_string_append(string, "Commands:\n");
+    for (gsize i = 0; i < context->entries_len; i++) {
+        entry = &context->entries[i];
+        entry_string = g_string_new(NULL);
+
+        g_string_append_printf(entry_string, "  %s", entry->name);
+        g_string_append_printf(entry_string, "%*s%s\n",
+                               (gint) (max_len - pu_utf8_strlen(entry_string->str)), "",
+                               entry->description ? entry->description : "");
+        g_string_append(string, g_string_free_and_steal(entry_string));
+    }
+
+    g_string_append_printf(string,
+                           "\nExecute any command with the '-h, --help' option "
+                           "to show command specific help.");
+
+    return g_string_free_and_steal(string);
+}
+
+static gchar *
+get_command_help(PuCommandContext *context,
+                 const gchar *command)
+{
+    GString *string = NULL;
+    PuCommandEntry *entry = NULL;
+
+    g_return_val_if_fail(context != NULL, NULL);
+    g_return_val_if_fail(command != NULL, NULL);
+
+    /* Find command in list of valid command entries */
+    for (gsize i = 0; i < context->entries_len; i++) {
+        if (g_str_equal(command, context->entries[i].name)) {
+            entry = &context->entries[i];
+            break;
+        }
+    }
+
+    if (!entry)
+        return NULL;
+
+    string = g_string_new(NULL);
+    g_string_append(string, entry->description);
```

Comment:
It does. Even better though, in this case, we could directly return `entry->description`, instead of constructing a `GString` with a single append and stealing its content.